sw/device/lib/dif/dif_aes.c

```diff
@@ -267,10 +286,64 @@ dif_result_t dif_aes_read_output(const dif_aes_t *aes, dif_aes_data_t *data) {
     return kDifError;
   }
 
-  for (int i = 0; i < AES_DATA_OUT_MULTIREG_COUNT; ++i) {
-    ptrdiff_t offset = AES_DATA_OUT_0_REG_OFFSET + (i * sizeof(uint32_t));
+  aes_read_multireg(aes, data->data, AES_DATA_OUT_MULTIREG_COUNT,
+                    AES_DATA_OUT_0_REG_OFFSET);
+
+  return kDifOk;
+}
+
+dif_result_t dif_aes_process_data(const dif_aes_t *aes,
+                                  const dif_aes_data_t *plain_text,
+                                  dif_aes_data_t *cipher_text,
+                                  size_t block_count) {
+  if (aes == NULL || plain_text == NULL || cipher_text == NULL ||
+      block_count == 0) {
+    return kDifBadArg;
+  }
+  const dif_aes_data_t *in_it = &plain_text[0];
+  const dif_aes_data_t *in_stop = &plain_text[block_count - 1];
+  dif_aes_data_t *out_it = &cipher_text[0];
+
+  // The algorithm below just makes sense for at least 2 blocks. Otherwise
+  // it is better to use the `load_data` and `read_output` functions.
+  if (block_count < 2) {
+    DIF_RETURN_IF_ERROR(dif_aes_load_data(aes, *in_it));
+    return dif_aes_read_output(aes, out_it);
+  }
+
+  // Ensure that the INPUT_READY bit in STATUS is 1.
+  if (!aes_input_ready(aes)) {
+    return kDifUnavailable;
+  }
+
+  // Write Input Data Block 0 to the Input Data registers DATA_IN_0 - DATA_IN_3.
+  aes_set_multireg(aes, in_it->data, AES_DATA_IN_MULTIREG_COUNT,
+                   AES_DATA_IN_0_REG_OFFSET);
+
+  // Wait for the INPUT_READY bit in STATUS to become 1, i.e. wait for the AES
+  // unit to load Input Data Block 0 into the internal state register and start
+  // operation.
+  AES_WAIT_FOR_STATUS(aes, AES_STATUS_INPUT_READY_BIT, true);
+  // Then for every Data Block I=0,..,N-3, software must:
+  for (int i = block_count; i > 0; --i) {
+    // Write Input Data Block I+1 into the Input Data register. There is no need
+    // to explicitly check INPUT_READY as in the same cycle OUTPUT_VALID becomes
+    // 1, the current input is loaded in (meaning INPUT_READY becomes 1 one
+    // cycle later).
+    if (++in_it <= in_stop) {
```

Comment:
Is there a reason it wouldn't be better to make this condition be `i < block_count - 1`, and then use `plain_text[i]->data` in the body?